doc/findproxyfile.1
.TH findproxyfile 1 "October 2004" "findproxyfile" "GridSite Manual"
.SH NAME
.B findproxyfile
\- returns full path to GSI Proxy file
.SH SYNOPSIS
.B findproxyfile 
[--proxycache=PATH] [--delegation-id=ID] [--user-dn=DN] [--outsidecache]
.SH DESCRIPTION
.B findproxyfile
returns full path to a GSI Proxy file, either in the proxy cache maintained
by the GridSite G-HTTPS and delegation portType functions, or in other
standard places.

If a User DN is given
.B findproxyfile 
uses the value of the 
.B --proxycache
argument, the GRST_PROXY_PATH or the
compile time default to determine the location of the proxy cache directory.
The directory is searched for a proxy having the given User DN and 
Delegation ID. (If no Delegation ID is specificed, then the default value is
used.)

If
.B findproxyfile
does not find a proxy or if a User DN is not given, but
.B --outsidecache
was given, then the environment variable X509_USER_PROXY and the standard
location /tmp/x509up_uUID are searched as well.

.SH OPTIONS

.IP "--proxycache=PATH"
Give the path of the proxy cache directory explicitly, overriding the
default and the GRST_PROXY_PATH environment variable if present.

.IP "--delegation-id=ID"
The optional delegation ID is search for in the proxy cache in addition to
the User DN. If absent, the default Delegation ID value is searched for.

.IP "--user-dn=DN"
The DN of the full user certificate associated with the proxy to be searched
for in the proxy cache. (This is not the DN of any proxy earlier in the
chain: it is a the DN of a certificate issued by a recognised CA.)

.IP "--outsidecache"
If a User DN is not given, or a proxy not found in the cache, then search
for a proxy using X509_USER_PROXY environment variable and file name of
form /tmp/x509up_uUID as well.

.SH RETURN VALUE
If a proxy is found, its full path is output on standard out.

.SH EXIT CODES
0 is returned on success. Non-zero otherwise.

.SH BUGS
In this version, no attempt is made to verify or validate the proxies.

.SH AUTHOR
Andrew McNab <Andrew.McNab@man.ac.uk>

findproxyfile is part of GridSite: http://www.gridsite.org/


CHANGES
- ==== GridSite version 2.3.5 ====
* Thu Sep 7 Mattias Ellert <mattias.ellert@physics.uu.se>
- Fix format compilation warnings
- Urlencode fix
* Wed Sep 27 Daniel Kouril <kouril@ics.muni.cz>
- Address some compiler warnings
* Fri Feb 2 František Dvořák <valtri@civ.zcu.cz>
- Address some warnings in htproxyput utility
* Tue Dec 11 František Dvořák <valtri@civ.zcu.cz>
- RPM packaging cleanups
- Update gridsite URL
- Change cgi scripts locations
- ==== GridSite version 2.3.4 ====
* Wed Sep 20 2017 Zdenek Sustr <sustr4@cesnet.cz>
- Fixing missing environment varialbles (Issue #35)
- ==== GridSite version 2.3.3 ====
* Thu Oct 25 2016 Zdenek Sustr <sustr4@cesnet.cz>
- Fixed an issue with parsing of robot certificates (GGUS #124499)
- ==== GridSite version 2.3.2 ====
* Thu Aug 16 2016 Zdenek Sustr <sustr4@cesnet.cz>
- Fixed stability issues in systemd-based distributions
- ==== GridSite version 2.3.1 ====
* Thu Aug 05 2016 Zdenek Sustr <sustr4@cesnet.cz>
- Assignment to NULL fixed, addressing gridsite module crashes
- ==== GridSite version 2.3.0 ====
* Thu Aug 03 2016 Zdenek Sustr <sustr4@cesnet.cz>
- A new GRST_VOMS_FQANS variable added
- Support for robot certificates added
- Manpage typo fixes
- Slash ('/') allowed in libdir
- Build procedure fixes
- Implicit declarations of isalnum() and isspace() fixed
- ==== GridSite version 2.2.6 ====
* Thu Jun 25 2015 Zdenek Sustr <sustr4@cesnet.cz>
- Fixing build with curl 7
- Linking to OpenSSL to avoid undefined symbols
- Fixing code that detects RFC proxies
- Fixing memory leaks (Fixes Issue #19, Issue #22)
- Thread-safe call OpenSSL_add_all_digests() (Fixes Issue #20)
- Adding missing pthread lib for Debian 8
- ==== GridSite version 2.2.5 ====
* Tue Jun 24 2014 Zdenek Sustr <sustr4@cesnet.cz>
- Fixing a memory handling issue in delegation code (Fixing GH Issue #16)
- ==== GridSite version 2.2.4 ====
* Tue Jun 3 2014 Zdenek Sustr <sustr4@cesnet.cz>
- Fixing several memory leaks
- Fixed handling of zero-length certificate arrays (Fixing GH Issue #15)
- ==== GridSite version 2.2.3 ====
* Tue Mar 25 2014 Zdenek Sustr <sustr4@cesnet.cz>
- Proper initialization of time constants in signature verification functions (GH Issue #11)
- ==== GridSite version 2.2.2 ====
* Wed Feb 26 2014 Zdenek Sustr <sustr4@cesnet.cz>
- Solving problems with overlapping delegation, issue #2
- ==== GridSite version 2.2.1 ====
* Fri Jan 17 2014 Zdenek Sustr <sustr4@cesnet.cz>
- Turning off TLS support explicitly in GridSite (previously unsupported in OpenSSL)
- ==== GridSite version 2.2.0 ====
* Mon Dec 2 2013 Zdenek Sustr <sustr4@cesnet.cz>
- Default key size for proxies is now 1024 rather than 512 b
- API is extended with functions, which allow developers to request custom key size (GH Issue #8)
- ==== GridSite version 2.1.3 ====
* Fri Aug 30 2013 Zdenek Sustr <sustr4@cesnet.cz>
- Maximum cert chain length determined dynamically, no hard-coded limit
- ==== GridSite version 2.1.2 ====
* Fri Aug 2 2013 Zdenek Sustr <sustr4@cesnet.cz>
- Proxy type set explicitly before calling the sign function
- ==== GridSite version 2.1.1 ====
* Fri Jul 19 2013 Zdenek Sustr <sustr4@cesnet.cz>
- Function GRST_is_id_safe() extended to accept dashes
- ==== GridSite version 2.1.0 ====
* Mon May 13 2013 Zdeněk Šustr <sustr4@cesnet.cz>
* New API function: GRST_is_id_safe()
* Fixed behavior when delegation operations overlap
* Fixed build and packaging issues in various distributions
- ==== GridSite version 2.0.4 ====
* Mon Jan 21 2013 Zdeněk Šustr <sustr4@cesnet.cz>
* When constructing a list of FQANs, a reference to the proxy is stored in the chain that contains the attribute (GGUS #79096)
* Avoid looking up remote IP, it is available in the Apache context
* Flapping yum update fixed
- ==== GridSite version 2.0.3 ====
* Wed Nov 14 2012 Zdeněk Šustr <sustr4@cesnet.cz>
- Segmentation fault in htproxyput fixed (occurred if run by non-root)
- ==== GridSite version 2.0.2 ====
* Tue Nov 06 2012 Zdeněk Šustr <sustr4@cesnet.cz>
- Certificates made available in the GRST structure
- ==== GridSite version 2.0.1 ====
* Fri Oct 22 2012 František Dvořák <valtri@civ.zcu.cz>
- one more update of the packaging for Debian
* Fri Oct 19 2012 František Dvořák <valtri@civ.zcu.cz>
- update of the packaging for Debian due to major version bump
- add DESTDIR to install target
* Thu Oct 18 2012 František Dvořák <valtri@civ.zcu.cz>
- fix packaging for SL6 and Fedora (curl-devel -> libcurl-devel)
- big library versions cleanup, using libtool to compile and link
- ==== GridSite version 2.0.0 ====
* Wed Oct 17 2012 Marcel Poul <marcel.poul@cern.ch>
- Internals rewritten to use caNl
* Wed Oct 17 2012 František Dvořák <valtri@civ.zcu.cz>
- Unused tools removed
- ==== GridSite version 1.7.29 ====
* Mon Dec 16 2013 Zdenek Sustr <sustr4@cesnet.cz>
- Default key length set to 1024 bits
- ==== GridSite version 1.7.28 ====
* Fri Aug 30 2013 Zdenek Sustr <sustr4@cesnet.cz>
- Maximum cert chain length determined dynamically, no hard-coded limit
- ==== GridSite version 1.7.27 ====
* Fri Jul 19 2013 Zdenek Sustr <sustr4@cesnet.cz>
- Function GRST_is_id_safe() extended to accept dashes
- ==== GridSite version 1.7.26 ====
* Fri May 24 2013 Zdenek Sustr <sustr4@cesnet.cz>
- New API function: GRST_is_id_safe()
- Fixed behavior when delegation operations overlap
- ==== GridSite version 1.7.25 ====
* Mon Dec 03 2012 Zdenek Sustr <sustr4@cesnet.cz>
- Allowing exceptions for length constraints in proxy certificates (GGUS #87573)
- ==== GridSite version 1.7.24 ====
* Tue Oct 30 2012 Zdenek Sustr <sustr4@cesnet.cz>
- Fixed Memory Leak
- ==== GridSite version 1.7.23 ====
* Tue Oct 09 2012 Zdenek Sustr <sustr4@cesnet.cz>
- Fixed Memory Leak
- Packaging fixes for SL6
- ==== GridSite version 1.7.22 ====
* Tue Jul 24 2012 František Dvořák <valtri@civ.zcu.cz>
- Proper obsoletes in rpm packages (versioned, not platform dependent)
- Dependency on pkg-config in devel package
- Provide pkg-config files for gridsite
- Create only source packages from make

src/findproxyfile.c
/*
   Copyright (c) 2002-6, Andrew McNab, University of Manchester
   All rights reserved.

   Redistribution and use in source and binary forms, with or
   without modification, are permitted provided that the following
   conditions are met:

     o Redistributions of source code must retain the above
       copyright notice, this list of conditions and the following
       disclaimer. 
     o Redistributions in binary form must reproduce the above
       copyright notice, this list of conditions and the following
       disclaimer in the documentation and/or other materials
       provided with the distribution. 

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
   BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
   TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
   ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
   OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
   OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
   POSSIBILITY OF SUCH DAMAGE.
*/

#ifndef VERSION
#define VERSION "0.0.0"
#endif

#define _GNU_SOURCE

#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <getopt.h>

#include "gridsite.h"
  
void printsyntax(char *argv0)
{
  char *p;

  p = rindex(argv0, '/');
  if (p != NULL) ++p;
  else           p = argv0;

  fprintf(stderr, "%s [--outsidecache] [--proxycache=PATH] "
                  "[--delegation-id=DELEGATION-ID] [--user-dn=USER-DN]\n"
                  "(Version: %s)\n", p, VERSION);
}
  
#define GRST_PROXY_CACHE "/var/www/proxycache"
  
int main(int argc, char *argv[])
{
  char  *delegation_id = "_", *proxycache = "", *user_dn = "",
        *proxyfile = NULL;        
  int    c, outsidecache = 0, verbose = 0, option_index;
  struct option long_options[] = {      {"verbose",     	0, 0, 'v'},
                                        {"outsidecache",	0, 0, 0},
                                        {"proxycache",		1, 0, 0},
                                        {"delegation-id",	1, 0, 0},
                                        {"user-dn",		1, 0, 0},
                                        {0, 0, 0, 0}  };

  if (argc == 1)
    {
      printsyntax(argv[0]);
      return 0;
    }

  while (1)
       {
         option_index = 0;
                                                                                
         c = getopt_long(argc, argv, "v", long_options, &option_index);

         if      (c == -1) break;
         else if (c == 0)
           {
             if      (option_index == 1) outsidecache  = 1;
             else if (option_index == 2) proxycache    = optarg;
             else if (option_index == 3) delegation_id = optarg;
             else if (option_index == 4) user_dn       = optarg;
           }
         else if (c == 'v') ++verbose;
       }
       
  if (*user_dn != '\0') /* try to find in proxy cache */
    {
      if ((proxycache == NULL) || (*proxycache == '\0'))
        proxycache = getenv("GRST_PROXY_CACHE");

      if ((proxycache == NULL) || (*proxycache == '\0'))
        proxycache = GRST_PROXY_CACHE;

      proxyfile = GRSTx509CachedProxyFind(proxycache, delegation_id, user_dn);
    }
    
  if (((proxyfile == NULL) || (*proxyfile == '\0')) && outsidecache)
    {
      proxyfile = GRSTx509FindProxyFileName();
    }

  if ((proxyfile != NULL) && (*proxyfile != '\0'))
    {
      puts(proxyfile);
      return 0;   
    }
    
  fputs("No proxy file found\n", stderr);
    
  return 1;
}
